File: src/easy/Q0389/Q0389.py
"""
389. Find the Difference
https://leetcode.com/problems/find-the-difference/
"""


# Dictionary Approach
# O(n) time, O(n) space, where n is the number of characters in s
def find_the_difference_1(s: str, t: str):
    count = {}
    for char in s:
        count[char] = count.get(char, 0) + 1
    for char in t:
        if count.get(char, 0) == 0:
            return char
        else:
            count[char] -= 1
    return ""


# Unicode Difference Approach
# O(n) time, O(1) space, where n is the number of characters in s
def find_the_difference_2(s: str, t: str):
    result = i = 0
    while i < len(s):
        result += ord(t[i]) - ord(s[i])
        i += 1
    result += ord(t[i])
    return chr(result)

File: src/main/cpp/easy/Q0389.cpp
// 389. Find the Difference
// https://leetcode.com/problems/find-the-difference/

#include <algorithm>
#include <string>
#include <unordered_map>

using namespace std;

// Time: O(n), Space: O(n), where n is the size of `t`.
char find_the_difference(string& s, string& t) {
  unordered_map<char, int> char_count;
  for (const char c : t) {
    if (char_count.find(c) == char_count.end()) {
      char_count[c] = 1;
    } else {
      char_count[c] += 1;
    }
  }
  for (const char c : s) {
    if (char_count.find(c) != char_count.end()) {
      char_count[c] -= 1;
    }
  }
  const auto op = [](pair<char, int> p) { return p.second > 0; };
  const auto result = find_if(char_count.begin(), char_count.end(), op);
  if (result == char_count.end()) {
    throw runtime_error("`s` and `t` were equal");
  }
  return (*result).first;
}
